### PR TITLE
compiler: Eliminate the Kernel Erlang intermediate representation

### PR DESCRIPTION
**Pull request by bjorng**
The undocumented Kernel Erlang intermediate format has been present in the compiler in some form since OTP R6B. In the current compiler, the `v3_kernel` pass translates Core Erlang to Kernel Erlang, and the `beam_kernel_to_ssa` pass then translates Kernel Erlang to SSA code.

Having to maintain the Kernel Erlang representation makes compiler maintenance harder. Therefore, this commit eliminates the Kernel Erlang representation by replacing the `v3_kernel` and `beam_kernel_to_ssa` passes with the `beam_core_to_ssa` pass.

While we are at it, we will guarantee that the SSA counter value in each function will never clash with an existing variable name. That will simplify generation of new variable names in the SSA passes.

For the history of Kernel Erlang and Core Erlang, see the following blog post:

https://www.erlang.org/blog/beam-compiler-history

### CODE REVIEW

File: lib/compiler/src/beam_ssa_pre_codegen.erl [hunk 1]
```diff
@@ -1526,7 +1526,7 @@ rce_reroute_terminator(#b_switch{list=List0}=Last, Exit, New) ->
 %%  in the exit block following the receive.
 
 recv_fix_common([Msg0|T], Exit, Rm, Blocks0, Count0) ->
-    {Msg,Count1} = new_var('@recv', Count0),
+    {Msg,Count1} = new_var('@ssa_recv', Count0),
```
Comment (jhogberg) [inline lib/compiler/src/beam_ssa_pre_codegen.erl:1529]: Do we still need this tag (same for `@retval`, `@ssa_stack` et al below)?

**Review by jhogberg — APPROVED** [review]

File: lib/compiler/src/beam_core_to_ssa.erl [hunk 2]
```diff
@@ -0,0 +1,3230 @@
...
+
+pattern_map_pairs(Ces0, Sub0, St0) ->
+    {Kes,{Sub1,St1}} =
+        mapfoldl(fun(#c_map_pair{key=Ck,val=Cv},{Subi0,Sti0}) ->
+                         {Kk,[],Sti1} = expr(Ck, Subi0, Sti0),
+                         {Kv,Subi2,Sti2} = pattern(Cv, Subi0, Sti1),
+                         {#cg_map_pair{key=Kk,val=Kv},{Subi2,Sti2}}
+                 end, {Sub0, St0}, Ces0),
+    %% It is later assumed that these keys are term sorted,
+    %% so we'll need to sort them here.
```
[the remark this answers is not in the document]
Comment (bjorng) [reply lib/compiler/src/beam_core_to_ssa.erl:806]: I think that has been handled earlier.

File: lib/compiler/src/beam_core_to_ssa.erl [hunk 3]
```diff
@@ -0,0 +1,3230 @@
...
+-record(kern, {module :: atom(),       %Current module
+               func,                   %Current host function
+               fargs=[] :: [#b_var{}], %Arguments for current function
+               vcount=0,               %Variable counter
+               fcount=0,               %Fun counter
+               ds=sets:new([{version, 2}]) :: sets:set(), %Defined variables
+               funs=[],                         %Fun functions
+               free=#{},                        %Free variables
+               ws=[]   :: [warning()],          %Warnings.
+               no_make_fun3 :: boolean(),
```
[the remark this answers is not in the document]
Comment (bjorng) [reply lib/compiler/src/beam_core_to_ssa.erl:151]: Yes, in a different PR.

File: lib/compiler/src/beam_ssa_pre_codegen.erl (hunk 1 above)
Comment (bjorng) [reply lib/compiler/src/beam_ssa_pre_codegen.erl]: I have now eliminated the tuple form of variable names, in order to further simplify the naming of variables. The only allowed allowed variable name is now a single atom or a single integer.

**Pull request merged**